common/map_test.cpp

```diff
@@ -0,0 +1,552 @@
+
+  EXPECT_FALSE(m.Erase(42));
+  EXPECT_TRUE(m.Contains(93));
+  EXPECT_TRUE(m.Erase(93));
+  EXPECT_FALSE(m.Contains(93));
+  for (int i : llvm::seq(102, 136)) {
+    SCOPED_TRACE(llvm::formatv("Key: {0}", i).str());
+    EXPECT_TRUE(m.Contains(i));
+    EXPECT_TRUE(m.Erase(i));
+    EXPECT_FALSE(m.Contains(i));
+  }
+  for (int i : llvm::seq(75, 175)) {
+    SCOPED_TRACE(llvm::formatv("Key: {0}", i).str());
+    if (i == 93 || (i >= 102 && i < 136)) {
+      continue;
+    }
+    ASSERT_TRUE(m.Contains(i));
+    EXPECT_EQ(i * 100 + 1, *m[i]);
+    EXPECT_FALSE(m.Insert(i, i * 100 + 2).is_inserted());
+    EXPECT_EQ(i * 100 + 1, *m[i]);
+    EXPECT_FALSE(m.Update(i, i * 100 + 2).is_inserted());
+    EXPECT_EQ(i * 100 + 2, *m[i]);
+  }
+  EXPECT_TRUE(m.Insert(93, 93 * 100 + 3).is_inserted());
+  EXPECT_EQ(93 * 100 + 3, *m[93]);
+
+  ExpectMapElementsAre(
+      m, MakeKeyValues([](int k) { return k * 100 + 2 + (k == 93); },
+                       llvm::seq(75, 102), llvm::seq(136, 175)));
+}
+
+template <typename MapT>
+class MapCollisionTest : public ::testing::Test {};
+
+using CollisionTypes = ::testing::Types<
+    Map<int, int, 16,
+        FixedHashKeyContext<7, /*FixIndexBits*/ true, /*FixTagBits*/ false, 0>>,
+    Map<int, int, 16,
+        FixedHashKeyContext<7, /*FixIndexBits*/ false, /*FixTagBits*/ true, 0>>,
+    Map<int, int, 16,
+        FixedHashKeyContext<7, /*FixIndexBits*/ true, /*FixTagBits*/ true, 0>>,
+    Map<int, int, 16,
+        FixedHashKeyContext<7, /*FixIndexBits*/ true, /*FixTagBits*/ true,
+                            ~static_cast<uint64_t>(0)>>>;
+TYPED_TEST_SUITE(MapCollisionTest, CollisionTypes);
+
+TYPED_TEST(MapCollisionTest, Basic) {
+  TypeParam m;
+
+  // Fill the map through a couple of growth steps, verifying at each step. Note
+  // that because this is a collision test, we synthesize actively harmful
+  // hashes in terms of collisions and so this test is essentially quadratic. We
+  // need to keep it relatively small.
+  for (int i : llvm::seq(1, 256)) {
+    SCOPED_TRACE(llvm::formatv("Key: {0}", i).str());
+    EXPECT_TRUE(m.Insert(i, i * 100).is_inserted());
+
+    // Immediately do a basic check of all elements to pin down when an
+    // insertion corrupts the rest of the table.
+    for (int j : llvm::seq(1, i)) {
+      SCOPED_TRACE(llvm::formatv("Assert key: {0}", j).str());
+      ASSERT_EQ(j * 100, *m[j]);
+    }
+  }
+  EXPECT_FALSE(m.Contains(257));
+
+  // Verify all the elements.
+  ExpectMapElementsAre(
+      m, MakeKeyValues([](int k) { return k * 100; }, llvm::seq(1, 256)));
+}
+
+TEST(MapContextTest, Basic) {
+  llvm::SmallVector<TestData> keys;
+  for (int i : llvm::seq(0, 513)) {
+    keys.push_back(i * 100);
+  }
+  IndexKeyContext<TestData> key_context(keys);
+  Map<ssize_t, int, 0, IndexKeyContext<TestData>> m;
+
+  EXPECT_FALSE(m.Contains(42, key_context));
+  EXPECT_TRUE(m.Insert(1, 100, key_context).is_inserted());
```

Comment:
Using * 100 for both the key and value here -- perhaps choose a different value or different value for `keys`?